pallets/identity/src/lib.rs

```diff
@@ -2226,42 +2226,65 @@ impl<T: Trait> CheckAccountCallPermissions<T::AccountId> for Module<T> {
         pallet_name: &PalletName,
         function_name: &DispatchableName,
     ) -> Option<AccountCallPermissionsData<T::AccountId>> {
+        // `who` is the original origin of the call (including forwarded calls from an identity acting as a key)
+        // context::current_identity is the target did against which the calling key must have permissions
+        // if `who`'s identity does not match context::current_identity we have a forwarded call, in this case the
+        // key for which we check permissions is `who`'s identity rather than `who`
         if <KeyToIdentityIds<T>>::contains_key(who) {
-            let primary_did = <KeyToIdentityIds<T>>::get(who);
-            if !<DidRecords<T>>::contains_key(&primary_did) {
+            let key_did = <KeyToIdentityIds<T>>::get(who);
+            if !<DidRecords<T>>::contains_key(&key_did) {
                 // The DID record is missing.
                 return None;
             }
-            let did_record = <DidRecords<T>>::get(&primary_did);
-            if who != &did_record.primary_key {
-                // `who` can be a secondary key.
-                //
-                // DIDs with frozen secondary keys (aka frozen DIDs) are not permitted to call
-                // extrinsics.
-                if Self::is_did_frozen(&primary_did) {
-                    // `primary_did` has its secondary keys frozen.
-                    return None;
+
+            if let Some(target_did) = Context::current_identity_or::<Self>(who).ok() {
+                let target_did_record = <DidRecords<T>>::get(&target_did);
+
+                if target_did == key_did {
+                    // In this case, we check the permissions of `who` in `target_did` as it is a direct call
+                    if who != &target_did_record.primary_key {
+                        // `who` can be a secondary key.
+                        //
+                        // DIDs with frozen secondary keys (aka frozen DIDs) are not permitted to call
+                        // extrinsics.
+                        if Self::is_did_frozen(&target_did) {
+                            // `target_did` has its secondary keys frozen.
+                            return None;
+                        }
+                        return target_did_record
+                            .secondary_keys
+                            .into_iter()
+                            .find(|sk| sk.signer == Signatory::Account(who.clone()))
+                            .filter(|sk| sk.has_extrinsic_permission(pallet_name, function_name))
+                            .map(|sk| AccountCallPermissionsData {
+                                primary_did: target_did,
+                                secondary_key: Some(sk),
+                            });
+                    }
+                } else {
+                    // In this case, we check the permissions of `who` in `target_did` as it is a direct call
```

Comment:
From the comment at the beginning of this function, I'd rather imagine
```suggestion
                    // In this case, we check the permissions of `who`'s identity in `target_did` as it is a forwarded call
```